Use nativeloader instead of system loader to load JNI library for soloader compatibility.

xcheng16 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29350 [Pytorch] system loader to native loader.**

Differential Revision: [D18352773](https://our.internmc.facebook.com/intern/diff/D18352773/)

Commit:
[Pytorch] system loader to native loader.



Differential Revision: [D18352773](https://our.internmc.facebook.com/intern/diff/D18352773/)

[ghstack-poisoned]

Commit:
Update on "[Pytorch] system loader to native loader."


Differential Revision: [D18352773](https://our.internmc.facebook.com/intern/diff/D18352773/)

[ghstack-poisoned]

Commit:
Update on "[Pytorch] system loader to native loader."


Differential Revision: [D18352773](https://our.internmc.facebook.com/intern/diff/D18352773/)

[ghstack-poisoned]

Commit:
Update on "[Pytorch] system loader to native loader."


Differential Revision: [D18352773](https://our.internmc.facebook.com/intern/diff/D18352773/)

[ghstack-poisoned]

dreiss commented on android/pytorch_android/CMakeLists.txt:
```diff
@@ -62,10 +62,8 @@ if (ANDROID_ABI)
-      -Wl,--whole-archive
+      -Wl,-dead_strip
       libtorch
-      -Wl,--no-whole-archive
```
This is for the Android build.  I don't think it needs to change.

dreiss commented on android/pytorch_android/CMakeLists.txt:
```diff
@@ -79,10 +77,8 @@ else()
-      -Wl,--whole-archive
+      -Wl,-dead_strip
       torch
-      -Wl,--no-whole-archive
```
Unfortunately, I suspect this breaks the Linux build, which links with gold.  Any way we can get both Mac and Linux builds to work?

Also, this is not related to the subject of the diff.  Please use separate diffs for unrelated changes.

dreiss commented on android/pytorch_android/src/androidTest/java/org/pytorch/PytorchHostTests.java:
```diff
@@ -1,28 +1,22 @@
 package org.pytorch;
 
-import com.facebook.soloader.nativeloader.NativeLoader;
```
It looks like this file was reformatted.  Can you do that as a separate diff first to make this change more self-contained?

dreiss commented on android/pytorch_android/src/androidTest/java/org/pytorch/PytorchHostTests.java:
```diff
@@ -1,28 +1,22 @@
-      NativeLoader.init(new SystemDelegate());
-    }
-  }
+  public static void setUpClass() {}
```
No longer needed.

dreiss commented on android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java:
```diff
@@ -1,30 +1,26 @@
-  public void setUp() {
-    System.loadLibrary("pytorch");
-  }
+  public void setUp() {}
```
No longer needed.

dreiss commented on android/pytorch_android/src/androidTest/java/org/pytorch/PytorchTestBase.java:
```diff
@@ -61,17 +57,17 @@ public void testEqFloat() throws IOException {
+          3.1415f,
+          -1,
+          0,
+          1,
```
Reformat should be a separate diff.

xcheng16 replied:
when update the diff, it auto-format the file, i can revert it.

dreiss commented on android/pytorch_android/src/main/cpp/pytorch_jni.cpp:
```diff
@@ -418,7 +418,7 @@ class JIValue : public facebook::jni::JavaClass<JIValue> {
       static const auto jMethodGetLong =
           JIValue::javaClassStatic()->getMethod<jlong()>("toLong");
-      return at::IValue{jMethodGetLong(jivalue)};
+      return at::IValue{(int64_t)jMethodGetLong(jivalue)};
```
This is part of the Mac PR, right?

xcheng16 replied:
yes, It is for mac, I will create separate pr for it.

dreiss reviewed (changes requested):
Nit: The "T" in PyTorch is capitalized.

Title shouldn't have "[PyTorch]" in GitHub (ghimport should automatically add it internally).

Title should probably clearly state "Use NativeLoader instead of system loader to load JNI library", and should say why we want to do this (so it's compatible with SoLoader).

Commit:
Update on "[Pytorch] system loader to native loader."


Differential Revision: [D18352773](https://our.internmc.facebook.com/intern/diff/D18352773/)

[ghstack-poisoned]
